proposals/1125_dynamic-supervisor-oidc-clients/README.md

```diff
@@ -0,0 +1,507 @@
+
+##### Client registry
+
+The supervisor's OIDC implementation currently performs live reads against the Kubernetes API (i.e. it does not use a
+cache).  No performance impact has been observed from this implementation.  A positive of this implementation is that
+the supervisor is always up to date with the latest state - i.e. if a session is deleted, it is immediately observed on
+the next API call that attempts to use that session.  The dynamic client registry must avoid using a cache based
+implementation to ensure that is always up to date with the current config.  Furthermore, the implementation must
+guarantee that deleting and recreating a client invalidates all sessions and client secrets associated with said client.
+Revocation of a client secret must invalidate all sessions that were authenticated using that client secret.
+
+##### Configuring association between clients and issuers
+
+There will be no configuration to associate a client with a particular issuer (i.e. federation domain).  Just as the
+`pinniped-cli` OAuth client is available for use with all federation domains, all dynamic clients will be available for
+use with all federation domains.
+
+#### Upgrades
+
+No backwards incompatible changes to any existing Kuberentes API resource schema are proposed in this design.
+
+The `pinniped-cli` client needs to continue to act as before for backwards compatibility with existing installations of
+the Pinniped CLI on user's machines. Therefore, it will temporarily be excluded from any new scope-based requirements
+which would restrict the username and group memberships from being returned.  When this exclusion is used, the
+supervisor will issue warnings and request that the user contact their admin for an updated kubeconfig.  This exclusion
+will be dropped in a future release.  Having this temporary exclusion will allow mixing of old CLIs with new
+supervisors, and new CLIs with old supervisors.  New CLIs will automatically include these new scopes in the kubeconfigs
+that they generate.
+
+#### Tests
+
+As usual, unit tests must be added for all new/changed code.
+
+Integration tests must be added to mimic the usage patterns of a webapp. Dynamic clients must be configured with
+various options to ensure that the options work as expected. Those clients must be used to perform the authcode flow,
+RFC8693 token exchanges, and TokenCredentialRequest calls. Negative tests must include validation failures on the new
+CRs, and failures to perform actions that are supposed to be disallowed on the client by its configuration. Integration
+tests will be used to mimic an upgrade scenario (with and without dynamic clients) to confirm that audience confusion is
+not possible.
+
+#### New Dependencies
+
+None are foreseen.
+
+#### Performance Considerations
+
+Extra calls will be made to the Kubernetes API to lookup dynamic OAuth clients.  No performance impact is foreseen.
```

Comment:
You mentioned the bcrypt performance consideration above, might be worth including here?

Reply:
Good point, thanks!

Reply:
Added.